# Water Project
**Clustering Data To Unveil the Water Crisis.**

## Introduction 
**Setting the stage for our data exploration journey.**

In [1]:
%load_ext sql

In [2]:
%sql mysql+pymysql://root:lilian80@localhost:3306/md_water_servicesds

'Connected: root@md_water_servicesds'

In [3]:
%%sql
SHOW TABLES;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
8 rows affected.


[('data_dictionary',),
 ('employee',),
 ('global_water_access',),
 ('location',),
 ('visits',),
 ('water_quality',),
 ('water_source',),
 ('well_pollution',)]

The steps used in this project are:
1. Cleaning the data.
2. Honouring the workers.
3. Analyzing locations.
4. Diving into the sources.
5. Starting a solution.
6. Analyzing queues.
7. Reporting insights.

## 1. Cleaning Our Data 
**Updating employee data.**

### 1.5. Add email Addresses
Assess the employee table. It has info on all of our workers, but note that the email addresses have not been added. Their emails must be updated since we will send them reports and figures. <br>
The emails should be in the format: `first_name.last_name@ndogowater.gov`. 

In [4]:
%%sql
SELECT * FROM employee Limit 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
5 rows affected.


[(0, 'Amara Jengo', '+99637993287', 'amara.jengo@ndogowater.gov', '36 Pwani Mchangani Road', 'Sokoto', 'Ilanga', 'Field Surveyor'),
 (1, 'Bello Azibo', '+99643864786', 'bello.azibo@ndogowater.gov', '129 Ziwa La Kioo Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (2, 'Bakari Iniko', '+99222599041', 'bakari.iniko@ndogowater.gov', '18 Mlima Tazama Avenue', 'Hawassa', 'Rural', 'Field Surveyor'),
 (3, 'Malachi Mavuso', '+99945849900', 'malachi.mavuso@ndogowater.gov', '100 Mogadishu Road', 'Akatsi', 'Lusaka', 'Field Surveyor'),
 (4, 'Cheche Buhle', '+99381679640', 'cheche.buhle@ndogowater.gov', '1 Savanna Street', 'Akatsi', 'Rural', 'Field Surveyor')]

In [5]:
%%sql
SELECT 
    employee_name,
    lower(REPLACE(employee_name, ' ', '.')) AS LowerFullName,
    CONCAT(lower(REPLACE(employee_name, ' ', '.')),'@ndogowater.gov') AS Email
FROM employee
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
5 rows affected.


[('Amara Jengo', 'amara.jengo', 'amara.jengo@ndogowater.gov'),
 ('Bello Azibo', 'bello.azibo', 'bello.azibo@ndogowater.gov'),
 ('Bakari Iniko', 'bakari.iniko', 'bakari.iniko@ndogowater.gov'),
 ('Malachi Mavuso', 'malachi.mavuso', 'malachi.mavuso@ndogowater.gov'),
 ('Cheche Buhle', 'cheche.buhle', 'cheche.buhle@ndogowater.gov')]

### 1.7. UPDATE email column
UPDATE the email column this time with the email addresses

In [6]:
%%sql 
UPDATE employee 
SET email = CONCAT(LOWER(REPLACE(employee_name, ' ', '.')),'@ndogowater.gov')

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
56 rows affected.


ResourceClosedError: This result object does not return rows. It has been closed automatically.

In [7]:
%%sql 
SELECT* FROM employee
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
5 rows affected.


[(0, 'Amara Jengo', '+99637993287', 'amara.jengo@ndogowater.gov', '36 Pwani Mchangani Road', 'Sokoto', 'Ilanga', 'Field Surveyor'),
 (1, 'Bello Azibo', '+99643864786', 'bello.azibo@ndogowater.gov', '129 Ziwa La Kioo Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (2, 'Bakari Iniko', '+99222599041', 'bakari.iniko@ndogowater.gov', '18 Mlima Tazama Avenue', 'Hawassa', 'Rural', 'Field Surveyor'),
 (3, 'Malachi Mavuso', '+99945849900', 'malachi.mavuso@ndogowater.gov', '100 Mogadishu Road', 'Akatsi', 'Lusaka', 'Field Surveyor'),
 (4, 'Cheche Buhle', '+99381679640', 'cheche.buhle@ndogowater.gov', '1 Savanna Street', 'Akatsi', 'Rural', 'Field Surveyor')]

In [8]:
%%sql
SELECT 
    LENGTH(phone_number) as PhoneLength,
    TRIM(phone_number) AS TrimmedPhone,
    LENGTH(TRIM(phone_number)) TrimmedLength
FROM employee
limit 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
5 rows affected.


[(12, '+99637993287', 12),
 (12, '+99643864786', 12),
 (12, '+99222599041', 12),
 (12, '+99945849900', 12),
 (12, '+99381679640', 12)]

In [9]:
%%sql
UPDATE employee 
SET phone_number = TRIM(phone_number)

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
56 rows affected.


ResourceClosedError: This result object does not return rows. It has been closed automatically.

In [10]:
%%sql 
SELECT 
    LENGTH (phone_number) as new_phone_length
FROM employee
LIMIT 3;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
3 rows affected.


[(12,), (12,), (12,)]

## 2. Honouring the workers.
**Finding our best.**

### 2.8. Employees in each town
Before diving into the analysis, let's have a look at where our employees live. <br>
Use the `employee` table to count how many of our employees live in each town.

In [11]:
%%sql
SELECT 
    town_name,
    COUNT(employee_name) as num_employees
FROM employee
GROUP BY town_name
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
5 rows affected.


[('Ilanga', 3), ('Rural', 29), ('Lusaka', 4), ('Zanzibar', 4), ('Dahabu', 6)]

### 10. Employees by locations visited 
Let's first look at the number of records each employee collected. Group, order, and limit the results to only see the top 3 `employee_ids` with the highest number of locations visited.

In [12]:
%%sql 
SELECT 
    assigned_employee_id,
    COUNT(visit_count) as number_of_visits
FROM visits
GROUP BY assigned_employee_id
ORDER BY 2 DESC
LIMIT 3;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
3 rows affected.


[(1, 3708), (30, 3676), (34, 3539)]

Using the top 3 `assigned_employee_id` to create a query that looks up the employee's info:` name`,` email address`, and` phone number` for the top employees.

In [13]:
%%sql 
SELECT 
    assigned_employee_id,
    employee_name,
    email,
    phone_number
FROM employee
WHERE assigned_employee_id IN (1,30,34);

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
3 rows affected.


[(1, 'Bello Azibo', 'bello.azibo@ndogowater.gov', '+99643864786'),
 (30, 'Pili Zola', 'pili.zola@ndogowater.gov', '+99822478933'),
 (34, 'Rudo Imani', 'rudo.imani@ndogowater.gov', '+99046972648')]

## 3. Analysing Locations
**Understanding where the water sources are.**

### 3.11. Number of records per town & province
From the `location` table, use the `province_name`, `town_name`, and `location_type` to understand where the water sources are by counting the number of records per town.

In [14]:
%%sql 
SELECT * FROM location LIMIT 3;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
3 rows affected.


[('AkHa00000', '2 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00001', '10 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00002', '9 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban')]

In [15]:
%%sql
SELECT 
    town_name,
    COUNT(town_name) as records_per_town
FROM location 
GROUP BY town_name 
ORDER BY 2 DESC
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
5 rows affected.


[('Rural', 23740),
 ('Harare', 1650),
 ('Amina', 1090),
 ('Lusaka', 1070),
 ('Mrembo', 990)]

Now count the records per province

In [16]:
%%sql
SELECT 
    province_name,
    COUNT(province_name) as records_per_province
FROM location 
GROUP BY province_name 
ORDER BY 2 DESC
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
5 rows affected.


[('Kilimani', 9510),
 ('Akatsi', 8940),
 ('Sokoto', 8220),
 ('Amanzi', 6950),
 ('Hawassa', 6030)]

### 3.12. Aggregate Functions
Finding the towns with the top 10 records in the different provinces.

In [6]:
%%sql
SELECT 
    province_name,
    town_name,
    COUNT(town_name) as records_per_town
FROM location
GROUP BY province_name, town_name
ORDER BY 3 DESC
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
10 rows affected.


[('Akatsi', 'Rural', 6290),
 ('Kilimani', 'Rural', 5440),
 ('Sokoto', 'Rural', 5010),
 ('Hawassa', 'Rural', 3900),
 ('Amanzi', 'Rural', 3100),
 ('Akatsi', 'Lusaka', 1070),
 ('Kilimani', 'Mrembo', 990),
 ('Amanzi', 'Asmara', 930),
 ('Amanzi', 'Dahabu', 930),
 ('Kilimani', 'Harare', 850)]

The number of records for each location type

In [18]:
%%sql
SELECT 
    location_type,
    COUNT(location_type) as num_sources
FROM location
GROUP BY location_type;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
2 rows affected.


[('Urban', 15910), ('Rural', 23740)]

In [19]:
%%sql
SELECT 
    ROUND(23740 / (15910 + 23740) * 100) as Rural_percentage,
    ROUND(15910 / (15910 + 23740) * 100) as Urban_percentage

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
1 rows affected.


[(Decimal('60'), Decimal('40'))]

## 4. Diving into the sources
**Seeing the scope of the problem.**

In [20]:
%%sql 
SELECT * FROM water_source LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
5 rows affected.


[('AkHa00000224', 'tap_in_home', 956),
 ('AkHa00001224', 'tap_in_home_broken', 930),
 ('AkHa00002224', 'tap_in_home_broken', 486),
 ('AkHa00003224', 'well', 364),
 ('AkHa00004224', 'tap_in_home_broken', 942)]

### 4.15 Exploring the `water_source` table
The way I look at this table; we have access to different water source types and the number of people using each source.
These are the questions that I am curious about.
1. How many people did we survey in total?
2. How many wells, taps and rivers are there?
3. How many people share particular types of water sources on average?
4. How many people are getting water from each type of source?

In [21]:
%%sql
# 1.how many people we surveyed in total
SELECT 
SUM(number_of_people_served) as total_people_served
FROM water_source;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
1 rows affected.


[(Decimal('27628140'),)]

In [22]:
%%sql
## 2.How many wells, taps, and rivers are there?
SELECT 
    type_of_water_source,
    COUNT(type_of_water_source) as number_of_sources
FROM water_source
GROUP BY type_of_water_source;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
5 rows affected.


[('tap_in_home', 7265),
 ('tap_in_home_broken', 5856),
 ('well', 17383),
 ('shared_tap', 5767),
 ('river', 3379)]

In [23]:
%%sql
# 3.How many people share particular types of water sources on average?
SELECT 
    type_of_water_source,
    ROUND(AVG(number_of_people_served)) as average_people_per_source
FROM water_source
GROUP BY type_of_water_source;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
5 rows affected.


[('tap_in_home', Decimal('644')),
 ('tap_in_home_broken', Decimal('649')),
 ('well', Decimal('279')),
 ('shared_tap', Decimal('2071')),
 ('river', Decimal('699'))]

These results are telling us that 644 people share a `tap_in_home` on average. 
<br>Does that make sense? No it doesn’t, right? 
<br>A few important things that apply to `tap_in_home` and `broken_tap_in_home` - The surveyors combined the data of many households together and added this as a single tap record, but each household actually has its own tap. In addition to this, there is an average of
6 people living in a home. So 6 people actually share 1 tap (not 644).This means that 1 `tap_in_home` actually represents 644 ÷ 6 = ± 100 taps.

In [24]:
%%sql
# 4. How many people are getting water from each type of source?
SELECT 
    type_of_water_source,
    SUM(number_of_people_served) as population_served
FROM water_source
GROUP BY type_of_water_source
ORDER BY 2 DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
5 rows affected.


[('shared_tap', Decimal('11945272')),
 ('well', Decimal('4841724')),
 ('tap_in_home', Decimal('4678880')),
 ('tap_in_home_broken', Decimal('3799720')),
 ('river', Decimal('2362544'))]

### 4.19. Percentage of People served per source 
It's a little hard to comprehend these numbers, but you can see that one of these is dominating. To make it a bit simpler to interpret, let's use
percentages. First, we need the total number of citizens then use the result of that and divide each of the` SUM(number_of_people_served)` by
that number, times 100, to get percentages.


In [25]:
%%sql
SELECT 
    type_of_water_source,
    ROUND(SUM(number_of_people_served)/27628140 * 100) as percentage_people_per_source
FROM water_source
GROUP BY type_of_water_source
ORDER BY 2 DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
5 rows affected.


[('shared_tap', Decimal('43')),
 ('well', Decimal('18')),
 ('tap_in_home', Decimal('17')),
 ('tap_in_home_broken', Decimal('14')),
 ('river', Decimal('9'))]

In [26]:
%%sql
SELECT 
    type_of_water_source,
    ROUND(SUM(number_of_people_served)/(
          SELECT 
          SUM(number_of_people_served)
          FROM water_source) * 100) as percentage_people_per_source
FROM water_source
GROUP BY type_of_water_source
ORDER BY 2 DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
5 rows affected.


[('shared_tap', Decimal('43')),
 ('well', Decimal('18')),
 ('tap_in_home', Decimal('17')),
 ('tap_in_home_broken', Decimal('14')),
 ('river', Decimal('9'))]

43% of our people are using shared taps in their communities, and on average, we saw earlier, that 2000 people share one `shared_tap`.
By adding `tap_in_home` and `tap_in_home_broken` together, we see that 31% of people have water infrastructure installed in their homes, but 45%
(14/31) of these taps are not working! This isn't the tap itself that is broken, but rather the infrastructure like treatment plants, reservoirs, pipes, and pumps that serve these homes that are broken.
18% of people are using `wells`. But only 4916 out of 17383 are clean = 28% (from last week).

## 5. Start a solution
**Thinking about how we can repair.**

### 5.23 Window Function

At some point, all the infrastructures will have be to fixed or improved, so we should start thinking about how we can make a data-driven decision and
how to do it. A simple approach is to fix the things that affect most people first. The following columns will be needed:
- Type of sources 
- Total people served grouped by the types
- A rank based on the total people served, grouped by the types

In [27]:
%%sql 
SELECT 
    type_of_water_source,
    SUM(number_of_people_served) as population_served,
    RANK() OVER(
        ORDER BY SUM(number_of_people_served) DESC) as rank_by_population
FROM water_source
GROUP BY type_of_water_source;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
5 rows affected.


[('shared_tap', Decimal('11945272'), 1),
 ('well', Decimal('4841724'), 2),
 ('tap_in_home', Decimal('4678880'), 3),
 ('tap_in_home_broken', Decimal('3799720'), 4),
 ('river', Decimal('2362544'), 5)]

So shared taps should fixed first, then wells, and so on. But the next question is, which shared taps or wells should be fixed first? <br>
Using the same logic; the most used sources should be fixed first.

### 5.24. Finding the Most Used Sources
The following steps could be used:
1. Assign a rank to the sources within each type.
2. Limit the results to only improvable sources.
3. Partition, filter, and order the results set.
4. Order the results to see the top of the list.

In [28]:
%%sql 
SELECT 
    source_id,
    type_of_water_source,
    number_of_people_served,
    RANK() OVER(
        PARTITION BY type_of_water_source ORDER BY number_of_people_served DESC) as priority_rank
FROM water_source
WHERE type_of_water_source NOT IN ('tap_in_home')
LIMIT 25;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
25 rows affected.


[('SoRu34798224', 'river', 998, 1),
 ('SoRu35837224', 'river', 998, 1),
 ('SoRu36238224', 'river', 998, 1),
 ('SoRu36791224', 'river', 998, 1),
 ('SoRu36880224', 'river', 998, 1),
 ('SoRu38142224', 'river', 998, 1),
 ('SoRu37756224', 'river', 998, 1),
 ('SoMa33775224', 'river', 998, 1),
 ('KiRu30353224', 'river', 998, 1),
 ('SoIl32972224', 'river', 998, 1),
 ('KiZu31236224', 'river', 998, 1),
 ('KiRu28591224', 'river', 998, 1),
 ('KiMr25030224', 'river', 998, 1),
 ('KiRu25603224', 'river', 996, 14),
 ('KiRu26027224', 'river', 996, 14),
 ('KiRu30532224', 'river', 996, 14),
 ('SoRu36229224', 'river', 996, 14),
 ('AkRu08053224', 'river', 996, 14),
 ('AmAb09070224', 'river', 996, 14),
 ('SoKo33179224', 'river', 994, 20),
 ('HaRu20469224', 'river', 994, 20),
 ('KiRu28404224', 'river', 994, 20),
 ('KiRu27661224', 'river', 994, 20),
 ('SoRu36506224', 'river', 994, 20),
 ('SoRu37912224', 'river', 994, 20)]

In [29]:
%%sql 
SELECT 
    source_id,
    type_of_water_source,
    number_of_people_served,
    DENSE_RANK() OVER(
        PARTITION BY type_of_water_source ORDER BY number_of_people_served DESC) as priority_rank
FROM water_source
WHERE type_of_water_source NOT IN ('tap_in_home')
LIMIT 30;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
30 rows affected.


[('SoRu34798224', 'river', 998, 1),
 ('SoRu35837224', 'river', 998, 1),
 ('SoRu36238224', 'river', 998, 1),
 ('SoRu36791224', 'river', 998, 1),
 ('SoRu36880224', 'river', 998, 1),
 ('SoRu38142224', 'river', 998, 1),
 ('SoRu37756224', 'river', 998, 1),
 ('SoMa33775224', 'river', 998, 1),
 ('KiRu30353224', 'river', 998, 1),
 ('SoIl32972224', 'river', 998, 1),
 ('KiZu31236224', 'river', 998, 1),
 ('KiRu28591224', 'river', 998, 1),
 ('KiMr25030224', 'river', 998, 1),
 ('KiRu25603224', 'river', 996, 2),
 ('KiRu26027224', 'river', 996, 2),
 ('KiRu30532224', 'river', 996, 2),
 ('SoRu36229224', 'river', 996, 2),
 ('AkRu08053224', 'river', 996, 2),
 ('AmAb09070224', 'river', 996, 2),
 ('SoKo33179224', 'river', 994, 3),
 ('HaRu20469224', 'river', 994, 3),
 ('KiRu28404224', 'river', 994, 3),
 ('KiRu27661224', 'river', 994, 3),
 ('SoRu36506224', 'river', 994, 3),
 ('SoRu37912224', 'river', 994, 3),
 ('SoRu35192224', 'river', 992, 4),
 ('SoRu36076224', 'river', 992, 4),
 ('SoRu37032224', 'river', 992, 4),
 ('SoMa33620224', 'river', 992, 4),
 ('AkRu04142224', 'river', 992, 4)]

In [30]:
%%sql 
SELECT 
    source_id,
    type_of_water_source,
    number_of_people_served,
    ROW_NUMBER() OVER(
        PARTITION BY type_of_water_source ORDER BY number_of_people_served DESC) as priority_rank
FROM water_source
WHERE type_of_water_source NOT IN ('tap_in_home')
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
10 rows affected.


[('SoRu34798224', 'river', 998, 1),
 ('SoRu35837224', 'river', 998, 2),
 ('SoRu36238224', 'river', 998, 3),
 ('SoRu36791224', 'river', 998, 4),
 ('SoRu36880224', 'river', 998, 5),
 ('SoRu38142224', 'river', 998, 6),
 ('SoRu37756224', 'river', 998, 7),
 ('SoMa33775224', 'river', 998, 8),
 ('KiRu30353224', 'river', 998, 9),
 ('SoIl32972224', 'river', 998, 10)]

## 6. Analysing queues
**Uncovering when citizens collect water.**

### 6.26. The `visits` Table
The `visits` table documented all of the visits our field surveyors made to each location. For most sources, one visit was enough, but if there were
queues, they visited the location a couple of times to get a good idea of the time it took for people to queue for water. So we have the time that
they collected the data, how many times the site was visited, and how long people had to queue for water.


These are some of the things worth looking at:
1. How long did the survey take?
2. What is the average total queue time for water?
<br>*Keeping in mind that many sources like taps_in_home have no queues. These are just recorded as 0 in the time_in_queue column, so when averages are calculated, those rows need to be excluded.*
3. What is the average queue time on different days?
4. How can we communicate this information efficiently?
<br>*Looking at `visits`, especially the `time_of_record` column. It is an SQL `DateTime` datatype, so the `DateTime` functions can be used to aggregate data for each day and even per hour*

In [31]:
%%sql 
SELECT * FROM visits LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
5 rows affected.


[(0, 'SoIl32582', 'SoIl32582224', datetime.datetime(2021, 1, 1, 9, 10), 1, 15, 12),
 (1, 'KiRu28935', 'KiRu28935224', datetime.datetime(2021, 1, 1, 9, 17), 1, 0, 46),
 (2, 'HaRu19752', 'HaRu19752224', datetime.datetime(2021, 1, 1, 9, 36), 1, 62, 40),
 (3, 'AkLu01628', 'AkLu01628224', datetime.datetime(2021, 1, 1, 9, 53), 1, 0, 1),
 (4, 'AkRu03357', 'AkRu03357224', datetime.datetime(2021, 1, 1, 10, 11), 1, 28, 14)]

In [32]:
%%sql
#1.How long did the survey take?
SELECT 
    MAX(time_of_record) as survey_finishing_date,
    MIN(time_of_record) as survey_starting_date,
    DATEDIFF(MAX(time_of_record), MIN(time_of_record)) as survey_time_taken
FROM visits;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
1 rows affected.


[(datetime.datetime(2023, 7, 14, 13, 53), datetime.datetime(2021, 1, 1, 9, 10), 924)]

In [33]:
%%sql
#2. What is the average total queue time for water?
SELECT 
    AVG(time_in_queue) as average_queue_time
FROM visits
WHERE time_in_queue <> 0;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
1 rows affected.


[(Decimal('123.2574'),)]

In [34]:
%%sql
# 3.What is the average queue time on different days?
SELECT
    DAYNAME(time_of_record) as day_of_week,
    ROUND(AVG(time_in_queue)) as avg_queue_time
FROM visits
GROUP BY DAYNAME(time_of_record);

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
7 rows affected.


[('Friday', Decimal('53')),
 ('Saturday', Decimal('246')),
 ('Sunday', Decimal('82')),
 ('Monday', Decimal('60')),
 ('Tuesday', Decimal('47')),
 ('Wednesday', Decimal('43')),
 ('Thursday', Decimal('46'))]

In [35]:
%%sql
# 4.How can we communicate this information efficiently?
SELECT 
    HOUR(time_of_record) as hour_of_day,
    ROUND(AVG(time_in_queue)) as avg_queue_time
FROM visits
GROUP BY HOUR(time_of_record)
ORDER BY HOUR(time_of_record) ASC;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
14 rows affected.


[(6, Decimal('149')),
 (7, Decimal('149')),
 (8, Decimal('149')),
 (9, Decimal('49')),
 (10, Decimal('48')),
 (11, Decimal('46')),
 (12, Decimal('47')),
 (13, Decimal('47')),
 (14, Decimal('47')),
 (15, Decimal('48')),
 (16, Decimal('47')),
 (17, Decimal('149')),
 (18, Decimal('147')),
 (19, Decimal('168'))]

Since the hour number is difficult to interpret, a format like 06:00 would be easier to read.

*To format time into a specific display format, the `TIME_FORMAT(time, format)` query can be used. It takes a time data field and
converts it into a format like %H:00 which is easy to read. `HOUR(time_of_record)` gives us an integer value of the hour of the day, that won't work
with `TIME_FORMAT()`, so `TIME(time_of_record)` needs to be used instead*.

In [36]:
%%sql
SELECT 
    TIME(time_of_record) as time
FROM visits
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
10 rows affected.


[(datetime.timedelta(seconds=33000),),
 (datetime.timedelta(seconds=33420),),
 (datetime.timedelta(seconds=34560),),
 (datetime.timedelta(seconds=35580),),
 (datetime.timedelta(seconds=36660),),
 (datetime.timedelta(seconds=37020),),
 (datetime.timedelta(seconds=37080),),
 (datetime.timedelta(seconds=37680),),
 (datetime.timedelta(seconds=38220),),
 (datetime.timedelta(seconds=39480),)]

In [37]:
%%sql
SELECT 
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
    ROUND(AVG(time_in_queue)) as avg_queue_time
FROM visits
GROUP BY 1
ORDER BY 1 ASC
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
5 rows affected.


[('06:00', Decimal('149')),
 ('07:00', Decimal('149')),
 ('08:00', Decimal('149')),
 ('09:00', Decimal('49')),
 ('10:00', Decimal('48'))]

### 6.32.Creating a pivot table in SQL
Pivot tables are not widely used in SQL, despite being useful for interpreting results. So there are no built-in functions to do this for us. Sometimes
the dataset is just so massive that it is the only option. *(For rows, the hour of the day will be used in that nice format, and then make each column a different day!)* 
<br>To filter a row `WHERE` is used, but using `CASE()` in `SELECT` can filter columns. `CASE()` function can be used for each day to separate the `queue time` column into a column for each day. Sunday can be used as an overview. So, when a row's `DAYNAME(time_of_record)` is Sunday, we
make that value equal to `time_in_queue`, and `NULL` for any days.

In [38]:
%%sql
SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
    DAYNAME(time_of_record) AS Day_name,
    CASE
        WHEN DAYNAME(time_of_record) = 'Sunday' 
        THEN time_in_queue
        ELSE NULL
    END AS Sunday
FROM visits
WHERE time_in_queue != 0 # This excludes other sources with 0 queue times.
LIMIT 30;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
30 rows affected.


[('09:00', 'Friday', None),
 ('09:00', 'Friday', None),
 ('10:00', 'Friday', None),
 ('10:00', 'Friday', None),
 ('11:00', 'Friday', None),
 ('11:00', 'Friday', None),
 ('12:00', 'Friday', None),
 ('12:00', 'Friday', None),
 ('13:00', 'Friday', None),
 ('13:00', 'Friday', None),
 ('13:00', 'Friday', None),
 ('14:00', 'Friday', None),
 ('14:00', 'Friday', None),
 ('15:00', 'Friday', None),
 ('16:00', 'Friday', None),
 ('07:00', 'Saturday', None),
 ('07:00', 'Saturday', None),
 ('09:00', 'Saturday', None),
 ('11:00', 'Saturday', None),
 ('13:00', 'Saturday', None),
 ('09:00', 'Sunday', 103),
 ('11:00', 'Sunday', 39),
 ('12:00', 'Sunday', 112),
 ('14:00', 'Sunday', 30),
 ('14:00', 'Sunday', 104),
 ('07:00', 'Monday', None),
 ('09:00', 'Monday', None),
 ('11:00', 'Monday', None),
 ('11:00', 'Monday', None),
 ('12:00', 'Monday', None)]

By adding `AVG()` around the `CASE()` function, the average is calculated, but since all of the other days' values are 0, we get an average for Sunday
only, rounded to 0 decimals. To aggregate by the hour, we can group the data by `hour_of_day`, and to make the table chronological, we also order
by `hour_of_day`.

In [39]:
%%sql
SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
# Sunday
    ROUND(AVG(
        CASE
            WHEN DAYNAME(time_of_record) = 'Sunday' 
            THEN time_in_queue ELSE NULL
        END 
    ),0) AS Sunday,
# Monday
    ROUND(AVG(
        CASE
            WHEN DAYNAME(time_of_record) = 'Monday' 
            THEN time_in_queue ELSE NULL
        END
    ),0) AS Monday,
# Tuesday
    ROUND(AVG(
        CASE
            WHEN DAYNAME(time_of_record) = 'Tuesday' 
            THEN time_in_queue ELSE NULL
        END
    ),0) as Tuesday,
# Wednesday
    ROUND(AVG(
        CASE
            WHEN DAYNAME(time_of_record) = 'Wednesday' 
            THEN time_in_queue ELSE NULL
        END
    ),0) as Wednesday,
# Thursday
    ROUND(AVG(
        CASE
            WHEN DAYNAME(time_of_record) = 'Thursday' 
            THEN time_in_queue ELSE NULL
        END
    ),0) as Thursday,
# Friday
    ROUND(AVG(
        CASE
            WHEN DAYNAME(time_of_record) = 'Friday' 
            THEN time_in_queue ELSE NULL
        END
    ),0) as Friday,
# Saturday
    ROUND(AVG(
        CASE
            WHEN DAYNAME(time_of_record) = 'Saturday' 
            THEN time_in_queue ELSE NULL
        END
    ),0) as Saturday
FROM visits
WHERE time_in_queue != 0 # This excludes other sources with 0 queue times
GROUP BY hour_of_day
ORDER BY hour_of_day;

 * mysql+pymysql://root:***@localhost:3306/md_water_servicesds
14 rows affected.


[('06:00', Decimal('79'), Decimal('190'), Decimal('134'), Decimal('112'), Decimal('134'), Decimal('153'), Decimal('247')),
 ('07:00', Decimal('82'), Decimal('186'), Decimal('128'), Decimal('111'), Decimal('139'), Decimal('156'), Decimal('247')),
 ('08:00', Decimal('86'), Decimal('183'), Decimal('130'), Decimal('119'), Decimal('129'), Decimal('153'), Decimal('247')),
 ('09:00', Decimal('84'), Decimal('127'), Decimal('105'), Decimal('94'), Decimal('99'), Decimal('107'), Decimal('252')),
 ('10:00', Decimal('83'), Decimal('119'), Decimal('99'), Decimal('89'), Decimal('95'), Decimal('112'), Decimal('259')),
 ('11:00', Decimal('78'), Decimal('115'), Decimal('102'), Decimal('86'), Decimal('99'), Decimal('104'), Decimal('236')),
 ('12:00', Decimal('78'), Decimal('115'), Decimal('97'), Decimal('88'), Decimal('96'), Decimal('109'), Decimal('239')),
 ('13:00', Decimal('81'), Decimal('122'), Decimal('97'), Decimal('98'), Decimal('101'), Decimal('115'), Decimal('242')),
 ('14:00', Decimal('83'), Decimal('127'), Decimal('104'), Decimal('92'), Decimal('96'), Decimal('110'), Decimal('244')),
 ('15:00', Decimal('83'), Decimal('126'), Decimal('104'), Decimal('88'), Decimal('92'), Decimal('110'), Decimal('248')),
 ('16:00', Decimal('83'), Decimal('127'), Decimal('99'), Decimal('90'), Decimal('99'), Decimal('109'), Decimal('251')),
 ('17:00', Decimal('79'), Decimal('181'), Decimal('135'), Decimal('121'), Decimal('129'), Decimal('151'), Decimal('251')),
 ('18:00', Decimal('80'), Decimal('174'), Decimal('122'), Decimal('113'), Decimal('132'), Decimal('158'), Decimal('240')),
 ('19:00', Decimal('127'), Decimal('159'), Decimal('145'), Decimal('176'), Decimal('137'), Decimal('103'), Decimal('282'))]

The following patterns can be spotted:
1. Queues are very long on a Monday morning and Monday evening as people rush to get water.
2. Wednesday has the lowest queue times, but long queues on Wednesday evening.
3. People have to queue pretty much twice as long on Saturdays compared to the weekdays. It looks like people spend their Saturdays queueing
for water, perhaps for the week's supply?
4. The shortest queues are on Sundays, and this is a cultural thing. The people prioritize family and religion, so Sundays are spent
with family and friends.

## 7. Reporting Insights 
**Assembling our insights to a story.**

### 7.38. Water Accessibility and Infrastructure Summary Report
This survey aimed to identify the water sources people use and determine both the total and average number of users for each source.
Additionally, it examined the duration citizens typically spend in queues to access water.

### 7.38 Insights
1. Most water sources are rural.
2. 43% of our people are using shared taps. 2000 people often share one tap.
3. 31% of our population has water infrastructure in their homes, but within that group, 45% face non-functional systems due to issues with pipes,
pumps, and reservoirs.
4. 18% of our people are using wells of which, but within that, only 28% are clean.
5. Our citizens often face long wait times for water, averaging more than 120 minutes.
6. In terms of queues:
- Queues are very long on Saturdays.
- Queues are longer in the mornings and evenings.
- Wednesdays and Sundays have the shortest queues.

### 7.39. Start of our plan
<br>A plan that could be implemented is:
1. To focus the efforts on improving the water sources that affect the most people.
- Most people will benefit if shared taps are improved first.
- Wells are a good source of water, but many are contaminated. Fixing this will benefit a lot of people.
- Fixing existing infrastructure will help many people. If they have running water again, they won't have to queue, thereby shorting queue times for
others. So two problems can be solved at once.
- Installing taps in homes will stretch our resources too thin, so for now if the queue times are low, that source won't be improved.
2. Most water sources are in rural areas. The teams need to know this as it means they will have to make these repairs/upgrades in
rural areas where road conditions, supplies, and labor are harder challenges to overcome.

### 7.40. Practical solutions
1. If communities are using **rivers**, trucks can be dispatched to those regions to provide water temporarily in the short term, while crews are sent out to drill for wells, providing a more permanent solution.
2. If communities are using **wells**, filters can be installed to purify the water. For wells with **biological** contamination, **UV filters** can be installed to kill microorganisms, and for *polluted wells*, **reverse osmosis** filters can be installed. In the long term, these sources need to be checked as to why they get polluted.
3. For **shared taps**, in the short term, additional water tankers can be sent to the busiest taps, on the busiest days. The queue time
pivot table we made could be used to send tankers at the busiest times. Meanwhile, **installation of extra taps** can be started where they are needed.
According to UN standards, the maximum acceptable wait time for water is 30 minutes. With this in mind, the aim is to **install taps** to get
**queue times below 30 min**.
4. **Shared taps** with **short queue times** (< 30 min) represent a logistical challenge to further reduce waiting times. The most effective solution,
installing taps in homes, is resource-intensive and better suited as a long-term goal.
5. **Addressing broken infrastructure** offers a significant impact even with just a single intervention. It is expensive to fix, but so **many people**
can **benefit** from repairing one facility. For example, fixing a reservoir or pipe that multiple taps are connected to. It is essential to find the
commonly affected areas to see where the problem actually is.